usr/src/uts/common/fs/zfs/dmu.c

```diff
@@ -1824,6 +1824,12 @@ dmu_sync_late_arrival(zio_t *pio, objset_t *os, dmu_sync_cb_t *done, zgd_t *zgd,
 	 */
 	zp->zp_nopwrite = B_FALSE;
 
+	/*
+	 * Record the vdev(s) backing this blkptr so they can be flushed when
+	 * the writes for the lwb have completed.
+	 */
+	zil_lwb_add_block(zgd->zgd_lwb, zgd->zgd_bp);
```

Comment:
I think the BP is allocated by zio_write(), so it is not valid at this point.

Reply:
Ah, I see. Based on that, we have to move these calls back into the `*_get_done` functions, right?

Reply:
I assume the goal is to make the zil_lwb_add_block() call from within the ZIL code rather than in each consumer (ZPL, ZVOL, ZTEST...).  So how about in `dmu_sync_done()`?

Reply:
Sure, that sounds reasonable. I think we'll have to add the `lwb_vdev_lock` back, too, since we won't be holding the `zl_issuer_lock` in those callbacks.

Reply:
Is the bp allocated by the zio_ready callback?  Looking though, it appears as if that executes synchronously when the zio is started from dmu_sync.  We would still hold the issuer lock at that point and the existing semantics would hold.

Reply:
If that were the case, moving the `zil_lwb_add_block()` call to `dmu_sync_read()` would cover our bases, I think.

Reply:
The BP will be allocated in the ready callback (and I *think* it can only be called once for ZIL blocks?  Not 100% on that).  However I don't think there's a guarantee that the ready callback is called from the same thread that calls zio_write().  In general the pipeline can call the ready callback later on, from a taskq.  That might not usually happen for ZIL writes, but it probably isn't something we want to rely on.

Reply:
In order to be able to replay dmu_sync blocks there needs to be a bp embedded in the log block data.   Right?   Shouldn't this allocation happen before zfs_get_data returns?

Reply:
If `dmu_sync()` (and `dmu_sync_late_arrival()`) were made to sleep until `dmu_sync_ready()` had been called, then we could hold true the invariant that the BP was allocated and accounted for in the lwb before `zl_get_data()` returned.  My initial dtracing indicated that it was the norm for that zio execution to occur on-thread for these ZIL writes, so the typical case would proceed unimpeded from a throughput perspective.

Reply:
I see.  Zio semantics ensure that children zios are ready before the parent can be ready.   Thanks for following that trail.  

I think there is a timing issue with  the zil_lwb_add_block() calls in dmu_sync_ready() when
1) defer_flush is true in zil_lwb_write_issue
2) there are dmu_sync blocks that are not yet ready
3) avl_swap will empty out the vdev_tree for the issuing lwb
4) the dmu_sync vdev will be added to the issuing lwb vdev_tree in dmu_sync_ready()
5) Something bad might happen?  (too many flushes or no flush depending if we flush with no waiters)  
  
And an even bigger window if the  zil_lwb_add_block() calls are in the dmu_sync_done() callback.